File: haskell/src/Chapter3Test.hs
import Test.Hspec
import Test.QuickCheck
import Prelude hiding(tail,
                      dropWhile,
                      takeWhile,
                      drop,
                      init,
                      length,
                      sum,
                      product,
                      reverse)
import Data.List (foldl')
import Chapter3 (List(Empty) ,
                 List(Cons),
                 tail,
                 setHead,
                 dropWhile,
                 takeWhile,
                 drop,
                 init,
                 foldRight,
                 length,
                 foldLeft,
                 sum,
                 product,
                 length2,
                 reverse,
                 foldLeftWithFoldRight,
                 foldRightWithFoldLeft,
                 append)

main = hspec $
 describe "Chapter3" $ do


  describe "tail" $ do
   it "works for empty List" $ do
    tail (Empty) `shouldBe` (Empty :: (Num a) => List a )
   it "should work for List with one element" $ do
    tail (Cons 1 Empty) `shouldBe` (Empty :: (Num a) => List a )
   it "should work for List with more than one element of type int" $ do
    tail (Cons 2 (Cons 1 Empty)) `shouldBe` (Cons 1 Empty)
   it "should work for List with more than one element of type String" $ do
    tail (Cons "2" (Cons "1" Empty)) `shouldBe` (Cons "1" Empty)
   it "should work for List of List with one element" $ do
    tail (Cons "2" Empty) `shouldBe` (Empty )
   it "should work for List of List" $ do
    tail (Cons (Cons "1" Empty) Empty) `shouldBe` (Empty)
   it "should work for List of List with many elements" $ do
    tail (Cons (Cons "2" Empty) (Cons (Cons "1" Empty) Empty) ) `shouldBe` (Cons (Cons "1" Empty) Empty )


  describe "setHead" $ do
   it "works for empty List" $ do
    setHead 1 (Empty) `shouldBe` (Cons 1 Empty)
   it "should work for List with one element" $ do
    setHead 2 (Cons 1 Empty) `shouldBe` (Cons 2 Empty)
   it "should work for List with more than one element" $ do
    setHead 3 (Cons 2 (Cons 1 Empty)) `shouldBe` (Cons 3 (Cons 1 Empty))

  describe "drop" $ do
   it "works for empty List" $ do
    drop 100 (Empty) `shouldBe` (Empty :: (Num a) => List a)
   it "should drop all elements for List when n > number of elements in List" $ do
    drop 100 (Cons 1 Empty) `shouldBe` (Empty)
   it "should drop 2 elements for List with 3 elements when n is 2" $ do
    drop 2 (Cons 3(Cons 2 (Cons 1 Empty))) `shouldBe` (Cons 1 Empty)
   it "should drop no element when n < 0" $ do
    drop (-1) (Cons 2 (Cons 1 Empty)) `shouldBe` (Cons 2 (Cons 1 Empty))


  describe "dropWhile" $ do
   it "works for empty List" $ do
    dropWhile (>0) (Empty) `shouldBe` (Empty)
   it "should drop all elements for List with one element satisfying the predicate" $ do
    dropWhile (>0) (Cons 1 Empty) `shouldBe` (Empty)
   it "should drop all elements for List with all elements satisfying the predicate" $ do
    dropWhile (>0) (Cons 2 (Cons 1 Empty)) `shouldBe` (Empty)
   it "should drop no element for List where no element satisfies the predicate" $ do
    dropWhile (<0) (Cons 2 (Cons 1 Empty)) `shouldBe` (Cons 2 (Cons 1 Empty))
   it "should drop 2 element for List of 4 where first 2 elements satisfies the predicate" $ do
    dropWhile (>0) (Cons 2 (Cons 1(Cons 0 (Cons (-1) Empty)))) `shouldBe` (Cons 0 (Cons (-1) Empty))
   it "should drop only 2 element for List of 6 where elements (0,1,4,5) satisfy the predicate but (2,3) don't" $ do
    dropWhile (>0) (Cons 2 (Cons 1(Cons 0 (Cons (-1) (Cons 2 (Cons 1 Empty)))))) `shouldBe` (Cons 0 (Cons (-1) (Cons 2 (Cons 1 Empty))))


  describe "takeWhile" $ do
   it "works for empty List" $ do
    takeWhile (>0) (Empty) `shouldBe` (Empty:: (Num a) => List a)
   it "should take all elements for List with one element satisfying the predicate" $ do
    takeWhile (>0) (Cons 1 Empty) `shouldBe` (Cons 1 Empty)
   it "should take all elements for List with all elements satisfying the predicate" $ do
    takeWhile (>0) (Cons 2 (Cons 1 Empty)) `shouldBe` (Cons 2 (Cons 1 Empty))
   it "should take no element for List where no element satisfies the predicate" $ do
    takeWhile (<0) (Cons 2 (Cons 1 Empty)) `shouldBe` (Empty)
   it "should take 2 element for List of 4 where first 2 elements satisfies the predicate" $ do
    takeWhile (>0) (Cons 2 (Cons 1(Cons 0 (Cons (-1) Empty)))) `shouldBe` (Cons 2 (Cons 1 Empty))
   it "should take only 2 element for List of 6 where elements (0,1,4,5) satisfy the predicate but (2,3) don't" $ do
    takeWhile (>0) (Cons 2 (Cons 1(Cons 0 (Cons (-1) (Cons 2 (Cons 1 Empty)))))) `shouldBe` (Cons 2 (Cons 1 Empty))

  describe "init" $ do
   it "works for empty List" $ do
    init (Empty) `shouldBe` (Empty :: (Num a) => List a)
   it "init on List with one element is Empty" $ do
    init (Cons 1 Empty) `shouldBe` (Empty)
   it "should take all elements except last for List with 2 elements" $ do
    init (Cons 2 (Cons 1 Empty)) `shouldBe` (Cons 2 Empty)
   it "should take all element except last for List with more than 2 elements" $ do
    init (Cons 3 (Cons 2 (Cons 1 Empty))) `shouldBe` (Cons 3(Cons 2 Empty))

  describe "foldRight" $ do
   it "works for empty List" $ do
    foldRight (+) 0 (Empty) `shouldBe` 0
   it "can sum a List with one element" $ do
    foldRight (+) 0 (Cons 1 Empty) `shouldBe` 1
   it "can sum a List with more than one element" $ do
    foldRight (+) 0 (Cons 1(Cons 2 Empty)) `shouldBe` 3
   it "can calculate a sum of a large List " $ do
    foldRight (+) 0 (foldr (\x t -> Cons x t) Empty [1..1000]  ) `shouldBe` 500500
   it "can calculate a product of a List with more than one element" $ do
    foldRight (*) 1 (Cons 3 (Cons 4(Cons 2 Empty))) `shouldBe` 24
   it "can calculate a product of a large List " $ do
    foldRight (*) 1 (foldr (\x t -> Cons x t) Empty [1..10]  ) `shouldBe` 3628800 -- 10!
   it "can short circuit a product of an infinite List if one element is 0" $ do
    foldRight (\x y -> if (x == 0) then 0 else x * y) 1 (foldr (\x t -> Cons x t) Empty [0,1..]) `shouldBe` 0

  describe "length with foldRight" $ do
   it "works for empty List" $ do
      length (Empty) `shouldBe` 0
   it "works for List with one element" $ do
      length (Cons 1 Empty) `shouldBe` 1
   it "works for List with many elements" $ do
      length (foldr (\x t -> Cons x t) Empty [1..10]  ) `shouldBe` 10


  describe "foldLeft" $ do
   it "works for empty List" $ do
      foldLeft (\t x -> x + t) (0) (Empty) `shouldBe` 0
   it "works for List with one element" $ do
      foldLeft (\t x -> x + t) (0) (Cons 1 Empty) `shouldBe` 1
   it "works for List with two elements" $ do
      foldLeft (\t x -> x + t) (0) (Cons 1 (Cons 2 Empty)) `shouldBe` 3
 -- foldLeft is strict
   it "works for large List with 100000 elements" $ do
     foldLeft (\t x -> x + t) (0) (foldl' (\t x -> Cons x t) Empty [1..100000]) `shouldBe` 5000050000


  describe "sum" $ do
   it "works for empty List" $ do
      sum  (Empty) `shouldBe` 0
   it "works for List with one element" $ do
      sum (Cons 1 Empty) `shouldBe` 1
   it "works for large List with 1000 elements" $ do
      sum (foldl' (\t x -> Cons x t) Empty [1..1000]) `shouldBe` 500500


  describe "product" $ do
   it "works for empty List" $ do
      product (Empty) `shouldBe` 1
   it "works for List with one element" $ do
      product (Cons 1 Empty) `shouldBe` 1
   it "works for large List with 10 elements" $ do
      product (foldl' (\t x -> Cons x t) Empty [1..10]) `shouldBe` 3628800


  describe "lengthWithFoldLeft" $ do
   it "works for empty List" $ do
      length2 (Empty) `shouldBe` 0
   it "works for List with one element" $ do
      length2 (Cons 1 Empty) `shouldBe` 1
   it "works for large List with 1000 elements" $ do
      length2 (foldl' (\t x -> Cons x t) Empty [1..1000]) `shouldBe` 1000


  describe "reverse" $ do
   it "works for empty List" $ do
      reverse (Empty :: (Num a) => List a) `shouldBe` Empty
   it "works for List with one element" $ do
      reverse (Cons 1 Empty) `shouldBe` (Cons 1 Empty)
   it "works for List with more than element" $ do
      reverse (Cons 1 (Cons 2 (Cons 3 Empty))) `shouldBe` (Cons 3 (Cons 2 (Cons 1 Empty)))


  describe "foldLeftWithFoldRight" $ do
   it "works for empty List" $ do
      foldLeftWithFoldRight (\t x -> x + t) 0 (Empty) `shouldBe` 0
   it "works for List with one element" $ do
      foldLeftWithFoldRight (\t x -> x + t) 0 (Cons 1 Empty) `shouldBe` 1
   it "works for large List with 1000 elements" $ do
      foldLeftWithFoldRight (\t x -> x + t) 0 (foldl' (\t x -> Cons x t) Empty [1..1000]) `shouldBe` 500500
   it "does reverses the list " $ do
      foldLeftWithFoldRight (\t x -> Cons x t) Empty (Cons 1 (Cons 2 (Cons 3 Empty))) `shouldBe` (Cons 3 (Cons 2 (Cons 1 Empty)))


  describe "foldRightWithFoldLeft" $ do
   it "works for empty List" $ do
      foldRightWithFoldLeft (\x t -> x + t) 0 (Empty) `shouldBe` 0
   it "works for List with one element" $ do
      foldRightWithFoldLeft (\x t -> x + t) 0 (Cons 1 Empty) `shouldBe` 1
   it "works for List with one element" $ do
      foldRightWithFoldLeft (\x t -> x - t) 0 (Cons 2 (Cons 3 (Cons 4 Empty))) `shouldBe` 3
   it "works for large List with 1000 elements" $ do
      foldRightWithFoldLeft (\x t -> x + t) 0 (foldl' (\t x -> Cons x t) Empty [1..1000]) `shouldBe` 500500
   it "does not reverse the list " $ do
      foldRightWithFoldLeft (\x t -> Cons x t) Empty (Cons 1 (Cons 2 (Cons 3 Empty))) `shouldBe` (Cons 1 (Cons 2 (Cons 3 Empty)))


  describe "append" $ do
   it "works for empty Lists" $ do
      append (Empty::(Num a)=> List a) (Empty::(Num a)=> List a) `shouldBe` (Empty::(Num a)=> List a)
   it "works when first list list is empty" $ do
      append (Empty) (Cons 1 Empty) `shouldBe` (Cons 1 Empty)
   it "works when second list is empty" $ do
      append (Cons 1 Empty) (Empty) `shouldBe` (Cons 1 Empty)
   it "works for one list with one element and other with many" $ do
      append (Cons 1 Empty) (Cons 3 ((Cons 2 Empty))) `shouldBe` (Cons 1 (Cons 3 (Cons 2 Empty)))
      append (Cons 3 ((Cons 2 Empty))) (Cons 1 Empty)  `shouldBe` (Cons 3 (Cons 2 (Cons 1 Empty)))
   it "works for larger Lists" $ do
      append (Cons 3 ((Cons 2 Empty))) (Cons 4 ((Cons 5 Empty)))  `shouldBe` (Cons 3 (Cons 2 (Cons 4 (Cons 5 Empty))))

File: scala/src/test/scala/com/ad/Chapter3Test.scala
package com.ad


import com.ad.Chapter3._

import org.scalatest._

class Chapter3Test extends FlatSpec with Matchers {

  // Exercise 3.2
  "tail" should "throw error on empty list" in {
    intercept[IllegalArgumentException] {
      tail(Nil)
    }
  }

  // Exercise 3.2
  "tail" should "return empty list for list with one element" in {
    tail(Cons(1, Nil)) should be(Nil)
  }

  // Exercise 3.2
  "tail" should "return 1 element list for list with two elements" in {
    tail(Cons(1, Cons(2, Nil))) should be(Cons(2, Nil))
  }

  // Exercise 3.3
  "setHead" should "empty list should set a new head" in {
    setHead(1, Nil) should be(Cons(1, Nil))
  }

  // Exercise 3.2
  "tail" should "return replace head of a list " in {
    setHead(1, Cons(2, Nil)) should be(Cons(1, Nil))
  }

  // Exercise 3.3 drop
  "drop" should " should drop one element for list with one element" in {
    drop(Cons(2, Nil), 1) should be(Nil)
  }

  "drop" should " should drop one element for list with more than one element" in {
    drop(Cons(2, Cons(3, Nil)), 1) should be(Cons(3, Nil))
  }


  "drop" should " should return empty list for empty list (drop empty list differs in behaviour to tail empty list) " in {
    drop(Nil, 1) should be(Nil)
  }

  "drop" should " return empty list when n is greater than length of the list" in {
    drop(Cons(2, Cons(3, Nil)), 10) should be(Nil)
  }

  // 3.4
  "tailWithDrop" should "throw exception for empty list (this is different to drop on empty list)" in {
    intercept[IllegalArgumentException] {
      tailWithDrop(Nil)
    }
  }
  // 3.4
  "tailWithDrop" should "return Nil for list with one element" in {
    tailWithDrop(Cons(1, Nil)) should be(Nil)
  }

  // 3.4
  "tailWithDrop" should "return List of one element for list with two elements" in {
    tailWithDrop(Cons(1, Cons(2, Nil))) should be(Cons(2, Nil))
  }

  // -- Exercise 3.5
  "dropWhile" should "return Nil for empty List" in {
    dropWhile(Nil, { x: Int => false }) should be(Nil)
  }

  // -- Exercise 3.5
  "dropWhile" should "return Nil where for all x => f(x) is true" in {
    dropWhile(Cons(2, Cons(3, Cons(4, Nil))), { x: Int => true }) should be(Nil)
  }

  // -- Exercise 3.5
  "dropWhile" should "return the same list where for all x => f(x) is false" in {
    dropWhile(Cons(2, Cons(3, Cons(4, Nil))), { x: Int => false }) should be(Cons(2, Cons(3, Cons(4, Nil))))
  }

  // -- Exercise 3.5
  "dropWhile" should "return a (Cons(2, Cons(0, Cons (4, Nil))) list when for (Cons 0, (Cons(2, Cons(3, Cons (4, Nil)))) for f(x) => x <= 0  " in {
    dropWhile(Cons(0, Cons(2, Cons(0, Cons(4, Nil)))), { x: Int => x <= 0 }) should be(Cons(2, Cons(0, Cons(4, Nil))))
  }

  // -- Exercise 3.6
  "init" should "return Nil for empty List" in {
    init(Nil) should be(Nil)
  }

  // -- Exercise 3.6
  "init" should "return Nil for List with one element" in {
    init(Cons(2, Nil)) should be(Nil)
  }

  // -- Exercise 3.6
  "init" should "return Cons 2 Nil for List (Cons(2, Cons(3, Nil)) " in {
    init(Cons(2, Cons(3, Nil))) should be(Cons(2, Nil))
  }

  // -- Exercise 3.6
  "takeWhile" should "return empty for empty list." in {
    takeWhile(Nil: List[Int])(x => x > 0) should be(Nil)
  }

  // -- Exercise 3.6
  "takeWhile" should "return empty for list with elements but on match." in {
    takeWhile(Cons(-2, Cons(-3, Nil)))(x => x > 0) should be(Nil)
  }

  // -- Exercise 3.6
  "takeWhile" should "return list with matching elements." in {
    takeWhile(Cons(-2, Cons(-3, Cons(4, Nil))))(x => x < 0) should be(Cons(-2, Cons(-3, Nil)))
  }

  // -- Exercise 3.6
  "takeWhile" should "stop taking more elements once it find a non matching element." in {
    takeWhile(Cons(-2, Cons(3, Cons(-4, Nil))))(x => x < 0) should be(Cons(-2, Nil))
  }

  // -- Exercise 3.7
  "foldRight" should " should compute sum correctly when folding function is sum." in {
    foldRight(Cons(-2, Cons(3, Cons(-4, Nil))))(0)((x, y) => x + y) should be(-3)
  }

  "foldRight" should " should compute difference correctly when folding function is minus." in {
    foldRight(Cons(-2, Cons(3, Cons(-4, Nil))))(0)((x, y) => x - y) should be(-9)
  }

  // -- Exercise 3.7
  "foldRight" should " with Cons function should just copy the list." in {
    foldRight(Cons(-2, Cons(3, Cons(-4, Nil))))(Nil: List[Int])((x, y) => Cons(x, y)) should be(Cons(-2, Cons(3, Cons(-4, Nil))))
  }

  // -- Exercise 3.7
  "foldRight" should " with Cons function and a Nil list should just return Nil." in {
    foldRight(Nil)(Nil: List[Int])((x, y) => Cons(x, y)) should be(Nil)
  }

  // Exercise 3.9
  "lengthWithFoldRight" should " work with an empty list." in {
    lengthWithFoldRight(Nil) should be(0)
  }

  // Exercise 3.9
  "lengthWithFoldRight" should " work with a non empty list." in {
    lengthWithFoldRight(Cons(2, Cons(3, Nil))) should be(2)
  }


  // -- Exercise 3.10
  "foldLeft" should " should compute sum correctly when folding function is sum." in {
    foldLeft(Cons(-2, Cons(3, Cons(-4, Nil))))(0)((x, y) => x + y) should be(-3)
  }

  "foldLeft" should " should compute difference correctly when folding function is minus." in {
    foldLeft(Cons(-2, Cons(3, Cons(-4, Nil))))(0)((t, x) => t - x) should be(3)
  }

  // -- Exercise 3.10
  "foldLeft" should " with Cons function should just reverse the list." in {
    foldLeft(Cons(-2, Cons(3, Cons(-4, Nil))))(Nil: List[Int])((t, x) => Cons(x, t)) should be(Cons(-4, Cons(3, Cons(-2, Nil))))
  }

  // -- Exercise 3.10
  "foldLeft" should " with Cons function and a Nil list should just return Nil." in {
    foldLeft(Nil)(Nil: List[Int])((x, y) => Cons(y, x)) should be(Nil)
  }

  // Exercise 3.11
  "sum" should " using fold should work correctly" in {
    sum(Cons(2, Cons(3, Cons(4, Nil)))) should be(9)
  }

  "sum" should " using fold on empty list should be 0" in {
    sum(Nil) should be(0)
  }

  "product" should " using fold should work correctly" in {
    product(Cons(2, Cons(-3, Cons(4, Nil)))) should be(-24)
  }

  "product" should " using fold on empty list should be 1" in {
    product(Nil) should be(1)
  }

  "length" should " using fold should work correctly" in {
    Chapter3.length(Cons(2, Cons(-3, Cons(4, Nil)))) should be(3)
  }

  "length" should " using fold on empty list should be 1" in {
    Chapter3.length(Nil) should be(0)
  }

  "reverse" should " using foldl on empty list should be empty" in {
    reverse(Nil) should be(Nil)
  }

  "reverse" should " using foldl should reverse the list" in {
    reverse(Cons(2, Cons(-3, Cons(4, Nil)))) should be(Cons(4, Cons(-3, Cons(2, Nil))))
  }


  //foldLeftWithFoldRight
  "foldLeftWithFoldRight" should " on empty list should be empty" in {
    foldLeftWithFoldRight(Nil)(Nil: List[Int])((t, x) => Cons(x, t)) should be(Nil)
  }

  "foldLeftWithFoldRight" should " using foldl should reverse the list" in {
    foldLeftWithFoldRight(Cons(-2, Cons(3, Cons(-4, Nil))))(Nil: List[Int])((t, x) => Cons(x, t)) should be(Cons(-4, Cons(3, Cons(-2, Nil))))
  }
}